### PR TITLE
Add documentation for SAML SP metadata changes in KC 21

### LINKED ISSUE #13606: Keycloak uses incorrect encryption keys as SAML identity brokers in SPSSODescriptor
### Describe the bug

Keycloak should send a SPSSODescriptor with a signature and an encryption key to a SAML identity provider. Instead of using a realm signature key for <md:KeyDescriptor use="signing"> and a realm encryption key for <md:KeyDescriptor use="encryption">, realm signing keys are used for signing AND encryption. Realm encryption keys are not considered at all.
I believe that the error is related to the following code location: 
https://github.com/keycloak/keycloak/blob/17.0.1/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
```
session.keys().getKeysStream(realm, KeyUse.SIG, Algorithm.RS256)
                    .filter(Objects::nonNull)
                    .filter(key -> key.getCertificate() != null)
                    .sorted(SamlService::compareKeys)
                    .forEach(key -> {
                        try {
                            Element element = SPMetadataDescriptor
                                    .buildKeyInfoElement(key.getKid(), PemUtils.encodeCertificate(key.getCertificate()));
                            signingKeys.add(element);

                            if (key.getStatus() == KeyStatus.ACTIVE) {
                                encryptionKeys.add(element);
                            }
                        } catch (ParserConfigurationException e) {
                            logger.warn("Failed to export SAML SP Metadata!", e);
                            throw new RuntimeException(e);
                        }
                    });
```
Instead of KeyUse.SIG probably KeyUse.ENC should be used.

### Version

19.0.1

### Expected behavior

Realm signing keys are used in the SPSSODescriptor only under KeyDescriptor use="signing" and realm encryption keys under KeyDescriptor use="encryption".

### Actual behavior

Currently in the SPSSODescriptor realm signing keys are used for signing AND encryption while realm encryption keys are not used at all.

### How to Reproduce?

Create an active realm rsa key for signing and another one for encryption. 
Add a SAML identity provider and call the SPSSODescriptor of this identity provider:
<host>/auth/realms/<realm-name>/broker/<identity-provider>/endpoint/descriptor

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

File: release_notes/topics/21_0_0.adoc [hunk 1]
```diff
@@ -10,4 +10,26 @@ See the migration guide for details.
...
 We removed the out-of-box support for Hashicorp vault in this release.
 
-See this https://github.com/keycloak/keycloak/discussions/16446[discussion] for more details.
\ No newline at end of file
+See this https://github.com/keycloak/keycloak/discussions/16446[discussion] for more details.
+
+= SAML SP metadata changes
+
+Prior to this release SAML SP metadata contained the same key for both
+signing and encryption use. Starting with this version of Keycloak
```
Comment (andymunro) [inline release_notes/topics/21_0_0.adoc]:
```suggestion
Prior to this release, SAML SP metadata contained the same key for both
signing and encryption use. Starting with this version of Keycloak,
```

File: release_notes/topics/21_0_0.adoc [hunk 2]
```diff
@@ -10,4 +10,26 @@ See the migration guide for details.
...
+we include only encryption intended realm keys for encryption use
+in SP metadata. For each encryption key descriptor we also specify
+the algorithm that it is supposed to be used with. The following table shows
+supported XML-Enc algorithms with the mapping to Keycloak realm keys.
```
Comment (andymunro) [inline release_notes/topics/21_0_0.adoc]:
```suggestion
the supported XML-Enc algorithms with the mapping to Keycloak realm keys.
```

File: server_admin/topics/identity-broker/saml.adoc [hunk 3]
```diff
@@ -64,8 +64,7 @@ We recommend to use some more secure algorithm instead of `*_SHA1`. Also, with `
 do not work if the SAML identity provider (for example another instance of {project_name}) runs on Java 17 or higher.
 
 |Encryption Algorithm
-|Encryption algorithm, which is used by SAML IDP for encryption of SAML documents, assertions or IDs. The corresponding decryption key
-for decrypt SAML document parts will be chosen based on this configured algorithm and should be available in realm keys for the encryption (ENC) usage.
+|Encryption algorithm, which is used by SAML IDP for encryption of SAML documents, assertions or IDs. The corresponding decryption key for decrypt SAML document parts will be chosen based on this configured algorithm and should be available in realm keys for the encryption (ENC) usage. If algorithm is not configured, then any supported algorithm is allowed and decryption key will be chosen based on the algorithm specified in SAML document itself.
```
Comment (andymunro) [inline server_admin/topics/identity-broker/saml.adoc]:
```suggestion
|Encryption algorithm, which is used by SAML IDP for encryption of SAML documents, assertions, or IDs. The corresponding decryption key for decrypt SAML document parts will be chosen based on this configured algorithm and should be available in realm keys for the encryption (ENC) usage. If the algorithm is not configured, any supported algorithm is allowed and a decryption key will be chosen based on the algorithm specified in SAML document itself.
```

File: upgrading/topics/keycloak/changes-21_0_0.adoc [hunk 4]
```diff
@@ -34,3 +34,18 @@ algorithms do not work on Java 17 or higher. If you use this algorithm and the o
...
+
+= SAML SP metadata changes
+
+In this version, Keycloak will refuse to decrypt assertions encrypted using
+a realm key generated for signing purpose. This change means all encrypted
+communication from IDP to SP (where Keycloak acts as the SP) will stop working.
+
+There are two ways to make this work:
+
+- either update IDP configuration with the metadata generated by a newer
```
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
- either update the IDP configuration with the metadata generated by a newer
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good. Just a few minor changes.

**Comment by mhajas** [comment]
Thank you for the review @andymunro! I updated the PR based on your suggestions.

**Review by andymunro — APPROVED** [review]
Thanks, @mhajas.  Great job.

**Review by mposolda — APPROVED** [review]

**Pull request merged**